### PR TITLE
Add attribute to signal that a binding supports function level retry

### PR DESCRIPTION
**Pull request by satvu**
### Issue describing the changes in this PR

#1442 and #1050 would require this change first. We do not want to generate retry policy info and populate function metadata incorrectly for types that do not support function level retry. [This table](https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-error-pages?tabs=fixed-delay%2Cin-process&pivots=programming-language-csharp#retries) shows which trigger types support function-level retries.

[Design doc](https://dev.azure.com/msazure/One/_git/AAPT-Antares-Docs?path=/TeamDocs/FunctionTeamDocs/Design/OOPWorkers/DotnetIsolated/TriggerCapabilities.md&_a=preview).

### Pull request checklist

* [x] My changes **do not** require documentation changes
  * [ ] Otherwise: Documentation issue linked to PR
* [x] My changes **should not** be added to the release notes for the next release
  * [x] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
  * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] I have added all required tests (Unit tests, E2E tests)

### CODE REVIEW

**Comment by satvu** [comment]
Blocking on design discussion topics noted in #1442

**Review by liliankasem — APPROVED** [review]

File: extensions/Worker.Extensions.Abstractions/src/BindingCapabilitiesAttribute.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+        /// <param name="capabilities">The list of capabilities</param>
+        public BindingCapabilitiesAttribute(params string[] capabilities)
+        {
+            Capabilities = capabilities;
+        }
+
+        /// <summary>
+        /// Lists the capabilities of a binding type.
+        /// </summary>
+        public string[]? Capabilities { get; private set; }
```
Comment (jviau) [inline extensions/Worker.Extensions.Abstractions/src/BindingCapabilitiesAttribute.cs:23]: consider falling back to `Array.Empty<string>()` - this lets callers avoid null checks.

File: extensions/Worker.Extensions.CosmosDB/src/CosmosDBTriggerAttribute.cs [hunk 2]
```diff
@@ -6,6 +6,7 @@
 
 namespace Microsoft.Azure.Functions.Worker
 {
+    [BindingCapabilities("FunctionLevelRetry")]
```
Comment (jviau) [inline extensions/Worker.Extensions.CosmosDB/src/CosmosDBTriggerAttribute.cs:9]: Can we introduce a public constants class to contain these capabilities?

``` CSHarp
public static class BindingCapabilities
{
    public const string FunctionLevelRetry = nameof(FunctionLevelRetry);
}
```

Although if this interferes with use `[BindingCapabilities()]` attribute due to name, then name the const class something else.

**Review by jviau — APPROVED** [review]

**Review by aishwaryabh — APPROVED** [review]

File: extensions/Worker.Extensions.Abstractions/src/BindingCapabilitiesAttribute.cs [hunk 3]
```diff
@@ -0,0 +1,25 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.Azure.Functions.Worker.Extensions.Abstractions
+{
+    [AttributeUsage(AttributeTargets.Class)]
+    public sealed class BindingCapabilitiesAttribute : Attribute
```
Comment (fabiocav) [inline extensions/Worker.Extensions.Abstractions/src/BindingCapabilitiesAttribute.cs:9]: Can you please add the XML documentation for the attribute?

**Review by fabiocav — CHANGES_REQUESTED** [review]
Minor comment, but two main questions about the PR:

1. Do we have an issue tracking this work? If so, can you please have it linked to the PR?
2. Can you please update the release notes with an appropriate description for the change?

File: extensions/Worker.Extensions.CosmosDB/src/CosmosDBTriggerAttribute.cs (hunk 2 above)
Comment (fabiocav) [reply extensions/Worker.Extensions.CosmosDB/src/CosmosDBTriggerAttribute.cs:9]: Would be good to have this as part of the changes, for the name, I actually like `KnownBindingCapabilities`, thoughts?

File: extensions/Worker.Extensions.Abstractions/src/BindingCapabilitiesAttribute.cs [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    /// <summary>
+    /// Specifies a binding type's capabilities. Intended for use on binding attribute classes only.
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Class)]
+    public sealed class BindingCapabilitiesAttribute : Attribute
+    {
+        /// <summary>
+        /// Initializes a new instance of <see cref="BindingCapabilitiesAttribute"/> with the specified list of capabilities.
+        /// </summary>
+        /// <param name="capabilities">The list of capabilities</param>
```
Comment (kshyju) [inline extensions/Worker.Extensions.Abstractions/src/BindingCapabilitiesAttribute.cs:17]: nit: missing . at the end of comment?

File: extensions/Worker.Extensions.Abstractions/src/KnownBindingCapabilities.cs [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+namespace Microsoft.Azure.Functions.Worker.Extensions.Abstractions
+{
+    /// <summary>
+    /// Class containing constant values representing known binding capabilities.
+    /// </summary>
+    public static class KnownBindingCapabilities
+    {
+        public const string FunctionLevelRetry = "FunctionLevelRetry";
```
Comment (kshyju) [inline extensions/Worker.Extensions.Abstractions/src/KnownBindingCapabilities.cs:11]: Do you want to add a summary comment to line 11?

File: extensions/Worker.Extensions.CosmosDB/release_notes.md [hunk 6]
```diff
@@ -3,3 +3,4 @@
 - My change description (#PR/#issue)
 -->
 - Fixed incorrect type of CosmosDBTriggerAttribute's `StartFromTime` property.
```
Comment (kshyju) [inline extensions/Worker.Extensions.CosmosDB/release_notes.md:5]: This [change was already released in 4.0.1](https://github.com/Azure/azure-functions-dotnet-worker/releases/tag/cosmosdb-extension-4.0.1). So remove it?

**Review by kshyju — APPROVED** [review]
Package version bump is missing from all impacted packages. Are you going to have another PR for that?

**Comment by satvu** [comment]
> Package version bump is missing from all impacted packages. Are you going to have another PR for that?

Bumped all but cosmos for now since that is in preview.

File: extensions/Worker.Extensions.Abstractions/release_notes.md [hunk 7]
```diff
@@ -0,0 +1,5 @@
+## Release notes
+<!-- Please add your release notes in the following format:
+- My change description (#PR/#issue)
+-->
```
Comment (fabiocav) [inline extensions/Worker.Extensions.Abstractions/release_notes.md:4]: Can you update the format here to match what's in the repo root? Would help ensure the release notes are consistent.

File: extensions/Worker.Extensions.CosmosDB/release_notes.md [hunk 8]
```diff
@@ -2,4 +2,5 @@
 <!-- Please add your release notes in the following format:
 - My change description (#PR/#issue)
 -->
-- Fixed incorrect type of CosmosDBTriggerAttribute's `StartFromTime` property.
+
+- Add `BindingCapabilities` attribute to CosmosDb trigger to express function-level retry capabilities.
```
Comment (fabiocav) [inline extensions/Worker.Extensions.CosmosDB/release_notes.md:6]: Same comment as above for the release notes.

File: extensions/Worker.Extensions.EventHubs/release_notes.md [hunk 9]
```diff
@@ -2,3 +2,4 @@
 <!-- Please add your release notes in the following format:
 - My change description (#PR/#issue)
 -->
+- Add `BindingCapabilities` attribute to EventHubTrigger to express function-level retry capabilities.
```
Comment (fabiocav) [inline extensions/Worker.Extensions.EventHubs/release_notes.md:5]: Same comment as above for the release notes.

File: extensions/Worker.Extensions.Kafka/release_notes.md [hunk 10]
```diff
@@ -2,8 +2,4 @@
 <!-- Please add your release notes in the following format:
 - My change description (#PR/#issue)
 -->
-
-### 3.7.0
-- Boosted the Kafka Performance for output binding (https://github.com/Azure/azure-functions-kafka-extension/pull/403)
-- Added the config linger.ms for increasing the timeout for output binding (https://github.com/Azure/azure-functions-kafka-extension/pull/407)
-- Fixing the compilation issue bug for nullability from attribute properties (https://github.com/Azure/azure-functions-kafka-extension/pull/362)
+- Add `BindingCapabilities` attribute to KafkaTrigger to express function-level retry capabilities.
```
Comment (fabiocav) [inline extensions/Worker.Extensions.Kafka/release_notes.md:5]: Same comment as above for the release notes.

File: extensions/Worker.Extensions.Timer/release_notes.md [hunk 11]
```diff
@@ -1,4 +1,5 @@
 ## Release notes
 <!-- Please add your release notes in the following format:
 - My change description (#PR/#issue)
--->
\ No newline at end of file
+-->
+- Add `BindingCapabilities` attribute to  TimerTrigger to express function-level retry capabilities.
```
Comment (fabiocav) [inline extensions/Worker.Extensions.Timer/release_notes.md:5]: Same comment as above for the release notes.

**Review by fabiocav — APPROVED** [review]

**Pull request merged**